csrc/host_ir/lower_to_llvm.cpp

```diff
@@ -0,0 +1,704 @@
+  for(auto* val : input_vals){
+    boundary_vals.insert(val);
+  }
+
+  IdModel id_model(&fusion);
+  const ValGraph& graph = id_model.buildExactGraph();
+  
+  for(long unsigned int i = 0; i < logical_domain.size(); i++){
+    auto* zero = builder.getInt64(i);
+    auto* input_i_ptr = builder.CreateGEP(int64Ty, input_ptr, zero, "ptr");
+    auto* input_i_val = builder.CreateLoad(int64Ty, input_i_ptr, "val");
+    val2stride[input_vals[i]] = StrideInfo();
+    val2stride[input_vals[i]].llvm_extent = input_i_val;
+  }
+
+  for(auto* val : output_vals){
+    auto index = mapToInputDomain(input_vals, val, graph);
+    if(index != -1){
+      val2stride[val] = val2stride[input_vals[index]];
+    }
+  }
+
+  llvm::Value* running_stride_product = builder.getInt64(1);
+  for(auto it = allocation_domain.rbegin(); it != allocation_domain.rend(); ++it){
+    generate_stride_llvm_ir(*it, val2stride, builder, boundary_vals, running_stride_product);
+  }
+
+  for(long unsigned int i = 0; i < logical_domain.size(); i++){
+    auto* output_i_ptr = builder.CreateGEP(int64Ty, output_ptr, builder.getInt64(i), "ptr");
+    builder.CreateStore(val2stride[input_vals[i]].llvm_stride, output_i_ptr);
+  }
+
+  builder.CreateRetVoid();
+  llvm::outs() << "=== LLVM IR ===\n";
+  Module->print(llvm::outs(), nullptr);
+  return llvm::orc::ThreadSafeModule(std::move(Module), std::move(Context));
+}
+
+
+/*
+
+Generate infer shape module
+
+*/
+llvm::orc::ThreadSafeModule generate_infer_shape_module(std::vector<IterDomain*>& input_logical_domain, std::vector<IterDomain*>& output_logical_domain, Fusion& fusion) {
+  auto Context = std::make_unique<llvm::LLVMContext>();
+  auto* ctx = Context.get();
+  auto Module = std::make_unique<llvm::Module>("infer_shape_module", *ctx);
+  llvm::IRBuilder<> builder(*ctx);
+  std::vector<llvm::Type*> output_types;
+  for(long unsigned int i = 0; i < output_logical_domain.size(); i++){
+    output_types.push_back(builder.getInt64Ty());
+  }
+  std::vector<llvm::Type*> input_types;
+  for(long unsigned int i = 0; i < input_logical_domain.size(); i++){
+    input_types.push_back(builder.getInt64Ty());
+  }
+  auto* int64Ty = llvm::Type::getInt64Ty(*ctx);
+  auto* ptrTy = llvm::PointerType::getUnqual(int64Ty);
+
+  auto* funcTy = llvm::FunctionType::get(llvm::Type::getVoidTy(*ctx), {ptrTy, int64Ty, ptrTy, int64Ty}, false);
+  auto* func = llvm::Function::Create(funcTy, llvm::Function::ExternalLinkage, "infer_shape", Module.get());
+  auto* entry = llvm::BasicBlock::Create(*ctx, "entry", func);
+  builder.SetInsertPoint(entry);
+  std::unordered_map<Val*, dependency_graph> val2graph;
+  std::vector<Val*> input_vals = domain2vals(input_logical_domain);
+  std::vector<Val*> output_vals = domain2vals(output_logical_domain);
+  auto arg_it = func->arg_begin();
+  llvm::Value* input_ptr = &*arg_it;
+  llvm::Value* output_ptr = &*arg_it+2;
+  IdModel id_model(&fusion);
+  const ValGraph& graph = id_model.buildExactGraph();
```

Comment:
we don't want to repetitively build the same id_model. Let's start thinking about a refactor so we can re-use this across a given fusion.